control-plane-operator/controllers/hostedapicache/cache.go

```diff
@@ -0,0 +1,170 @@
+	}
+	h.cache = nil
+	h.cancelCacheCtx = nil
+	h.kubeConfig = nil
+}
+
+func (h *hostedAPICache) update(requestCtx context.Context, relatedObj client.Object, newKubeConfig []byte) error {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	if len(newKubeConfig) == 0 {
+		return fmt.Errorf("kube config is empty")
+	}
+
+	// Only rebuild the cache if the kubeconfig has changed
+	if bytes.Equal(newKubeConfig, h.kubeConfig) {
+		return nil
+	}
+
+	h.log.Info("rebuilding api cache")
+
+	// Initialize a new cache
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(newKubeConfig)
+	if err != nil {
+		return fmt.Errorf("invalid kube config: %w", err)
+	}
+	newCache, err := cache.New(restConfig, cache.Options{
+		Scheme: h.scheme,
+		Mapper: h.mapper,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create cache: %w", err)
+	}
+
+	// Initialize event handlers for the new cache
+	clusterOperatorInformer, err := newCache.GetInformerForKind(requestCtx, schema.GroupVersionKind{
+		Group:   configv1.GroupName,
+		Version: configv1.GroupVersion.Version,
+		Kind:    "ClusterOperator",
+	})
+	if err != nil {
+		return fmt.Errorf("failed to set up cache informers: %w", err)
+	}
+	clusterOperatorInformer.AddEventHandler(toolscache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			h.log.Info("DEBUG: adding genericevent for add")
+			h.events <- event.GenericEvent{Object: relatedObj}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldClusterOperator := oldObj.(*configv1.ClusterOperator)
+			newClusterOperator := newObj.(*configv1.ClusterOperator)
+			if !equality.Semantic.DeepEqual(oldClusterOperator.Status.Conditions, newClusterOperator.Status.Conditions) {
+				h.log.Info("DEBUG: adding genericevent for update")
+				h.events <- event.GenericEvent{Object: relatedObj}
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			h.log.Info("DEBUG: adding genericevent for delete")
+			h.events <- event.GenericEvent{Object: relatedObj}
+		},
+	})
+
+	// Shut down any existing cache
+	if h.cancelCacheCtx != nil {
+		h.cancelCacheCtx()
+	}
+
+	// Replace the existing cache
+	newCacheCtx, newCancelCache := context.WithCancel(h.ctx)
+	h.cache = newCache
+	h.cancelCacheCtx = newCancelCache
+	h.kubeConfig = newKubeConfig
+
+	// Start the new cache
+	go func() {
+		if err := h.cache.Start(newCacheCtx); err != nil {
+			h.log.Error(err, "failed to start hosted api cache")
+		} else {
+			h.log.Info("hosted api cache gracefully stopped")
+		}
+	}()
+
+	h.log.Info("successfully rebuilt api cache")
+
+	return nil
+}
+
+func (h *hostedAPICache) Get(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+	h.lock.Lock()
+	if h.cache == nil {
+		return NotInitializedError
+	}
+	h.lock.Unlock()
+	return h.cache.Get(ctx, key, obj)
+}
+
+func (h *hostedAPICache) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+	h.lock.Lock()
+	if h.cache == nil {
+		return NotInitializedError
+	}
+	h.lock.Unlock()
```

Comment:
If h.cache is nil, then you remain locked, is that on purpose? shouldn't you defer Unlock()?

Reply:
Good catch — I was deferring and then decided not to hold the lock for the cache call itself and failed to unlock before the return. Probably less error prone to just keep the lock for the duration of the call given we have no serious performance considerations here yet. Thanks!